snapshot/storage/ql/ql.go

```diff
@@ -0,0 +1,358 @@
+package ql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"path/filepath"
+
+	"github.com/docker/containerd/snapshot"
+	"github.com/docker/containerd/snapshot/storage"
+	"github.com/pkg/errors"
+
+	_ "github.com/cznic/ql/driver"
+)
+
+type qlMetastore struct {
+	dbfile string
+}
+
+// NewMetaStore returns a snapshot MetaStore for storage of metadata related to
+// a snapshot driver backed by a ql file database. This implementation is
+// strongly consistent and does all metadata changes in a transaction to prevent
+// against process crashes causing inconsistent metadata state.
+//
+// To view mtastore with a utility use https://godoc.org/github.com/cznic/ql/ql
+func NewMetaStore(ctx context.Context, dbfile string) (storage.MetaStore, error) {
+	ms := &qlMetastore{
+		dbfile: dbfile,
+	}
+	schema := []string{
+		`CREATE TABLE IF NOT EXISTS Snapshots (Name string NOT NULL, Parent int, Active bool, Readonly bool);`,
```

Comment:
How come `Parent` is an int and not a string?

I'm wondering if it might be clearer to have `Active` instead be an enumerated type `SnapshotType`.

It'd likely also be a good idea to include a few more things here:
* Creation timestamp
* Update timestamp
* Deletion (toombstone) timestamp

(Toombstoning/soft deleting can be a good thing to build in from the start, and then handle hard deletion/garbage collection separately.  This will enable the database to be repaired/recovered by an operator if something goes sideways.)